README.md
### Tags and respective `Dockerfile` links

- [`2.3.1`](https://github.com/jdeathe/centos-ssh-haproxy/releases/tag/2.3.1), `centos-7` [(centos-7/Dockerfile)](https://github.com/jdeathe/centos-ssh-haproxy/blob/centos-7/Dockerfile)
- [`1.3.1`](https://github.com/jdeathe/centos-ssh-haproxy/releases/tag/1.3.1), `centos-6` [(centos-6/Dockerfile)](https://github.com/jdeathe/centos-ssh-haproxy/blob/centos-6/Dockerfile)

## Overview

This build uses the base image [jdeathe/centos-ssh](https://github.com/jdeathe/centos-ssh) so inherits it's features but with `sshd` disabled by default. [Supervisor](http://supervisord.org/) is used to start the [haproxy](http://www.haproxy.org/) daemon when a docker container based on this image is run.

To manage HAProxy both [HATop](http://feurix.org/projects/hatop/) and [socat](http://www.dest-unreach.org/socat/) are included to provide both a UI and low level cli management interface respectively.

### Image variants

- [HAProxy 1.8 / HATop 0.7 - CentOS-7](https://github.com/jdeathe/centos-ssh-haproxy/blob/centos-7)
- [HAProxy 1.5 / HATop 0.7 - CentOS-6](https://github.com/jdeathe/centos-ssh-haproxy/blob/centos-6)

## Quick start

> For production use, it is recommended to select a specific release tag as shown in the examples.

Run up a container named `haproxy.1` from the docker image `jdeathe/centos-ssh-haproxy` on port 80 and 443 of your docker host. 2 backend hosts, `httpd_1` and `httpd_2`, are defined with IP addresses `172.17.8.101` and `172.17.8.102`; this is required to identify the backend hosts from within the HAProxy configuration file if not using docker network aliases.

> Change `172.17.8.101` and `172.17.8.102` in the example below to IP addresses that resolve to valid web servers on your network.

```
$ docker run -d -t \
  --name haproxy.1 \
  -p 80:80 \
  -p 443:443 \
  --add-host httpd_1:172.17.8.101 \
  --add-host httpd_2:172.17.8.102 \
  jdeathe/centos-ssh-haproxy:2.3.1
```

Verify the named container's process status and health.

```
$ docker ps -a \
  -f "name=haproxy.1"
```

Verify successful initialisation of the named container.

```
$ docker logs haproxy.1
```

## Instructions

### Running

To run the a docker container from this image you can use the standard docker commands. Alternatively, there's a [docker-compose.yml](https://github.com/jdeathe/centos-ssh-haproxy/blob/centos-7/docker-compose.yml) example.

In the following example the http service is bound to port 80 and https on port 443 of the docker host. Also, the environment variable `HAPROXY_HOST_NAMES` has been used to set a list of 3 hostnames to be added to the auto-generated self-signed SAN certificate.

#### Using environment variables

```
$ docker stop haproxy.1 && \
  docker rm haproxy.1; \
  docker run \
  --detach \
  --tty \
  --name haproxy.1 \
  --publish 80:80 \
  --publish 443:443 \
  --sysctl "net.core.somaxconn=32768" \
  --sysctl "net.ipv4.ip_local_port_range=1024 65535" \
  --sysctl "net.ipv4.route.flush=1" \
  --ulimit memlock=82000 \
  --ulimit nofile=131072 \
  --ulimit nproc=65535 \
  --env "HAPROXY_HOST_NAMES=www.app.local app.local localhost.localdomain" \
  --add-host httpd_1:172.17.8.101 \
  --add-host httpd_2:172.17.8.102 \
  jdeathe/centos-ssh-haproxy:2.3.1
```

#### Environment variables

There are several environmental variables defined at runtime which allows the operator to customise the running container.

##### HAPROXY_SSL_CERTIFICATE

The HAProxy SSL/TLS certificate can be defined using `HAPROXY_SSL_CERTIFICATE`. The value may be either a file path, a base64 encoded string of the certificate file contents or a multiline string containing a PEM formatted concatenation of private key and certificate. If set to a file path the contents may also be a base64 encoded string.

> **Note:** If using a file path with base64 encoded content the on container path is not maintained so this feature is *not* suitable for use with orchestration secrets such as Docker Swarm secrets or config however use of unencoded file contents is.

##### HAPROXY_CONFIG

The HAProxy configuration file path, (or base64 encoded string of the configuration file contents), is set using `HAPROXY_CONFIG`. The default http configuration is located at the path `/etc/haproxy/haproxy-http.example.cfg` and will be copied into the running configuration path `/etc/haproxy/haproxy.cfg`. There's also an example tcp configuration in the path `/etc/haproxy/haproxy-tcp.example.cfg`.

> **Note:** In most situations it will be necessary to define a custom configuration where the use of the base64 encoded string option is recommended. However if using a file path with base64 encoded content the on container path is not maintained so this feature is *not* suitable for use with orchestration secrets such as Docker Swarm secrets or config so when using these unencoded file contents is recommended.

##### HAPROXY_HOST_NAMES

The `HAPROXY_HOST_NAMES` can be used to set a, space separated, list of hostnames to be added to the automatically generated self-signed SAN certificate.

> **Note:** The value is unused if `HAPROXY_SSL_CERTIFICATE` has a valid value set.


CHANGELOG.md
# Change Log

## 2 - centos-7

Summary of release changes.

### 2.3.1 - 2019-10-10

- Deprecate Makefile target `logs-delayed`; replaced with `logsdef`.
- Updates `rsyslog` package to 8.24.0-41.el7_7.
- Updates source image to [2.6.1](https://github.com/jdeathe/centos-ssh/releases/tag/2.6.1).
- Updates `test/health_status` helper script with for consistency.
- Updates Makefile target `logs` to accept `[OPTIONS]` (e.g `make -- logs -ft`).
- Updates `healthcheck` script; state file existence confirms bootstrap completion.
- Updates healthcheck failure messages to remove EOL character that is rendered in status response.
- Updates wrapper script; only emit "waiting on" info message if bootstrap hasn't completed.
- Updates ordering of Tags and respective Dockerfile links in README.md for readability.
- Adds improved test workflow; added `test-setup` target to Makefile.
- Adds Makefile target `logsdef` to handle deferred logs output within a target chain.
- Adds `/docs` directory for supplementary documentation and simplify README.
- Fixes docker-compose example configuration files to work on CentOS-7 hosts.
- Fixes validation failure of 0 second --timeout value in `test/health_status`.

### 2.3.0 - 2019-08-20

- Updates `haproxy18u` packages to 1.8.20-1.
- Updates source image to [2.6.0](https://github.com/jdeathe/centos-ssh/releases/tag/2.6.0).
- Updates CHANGELOG.md to simplify maintenance.
- Updates README.md to simplify contents and improve readability.
- Updates README-short.txt to apply to all image variants.
- Updates Dockerfile `org.deathe.description` metadata LABEL for consistency.
- Updates supervisord configuration to send error log output to stderr.
- Updates haproxy-bootstrap supervisord configuration file/priority to `20-haproxy-bootstrap.conf`/`20`.
- Updates resyslogd-wrapper supervisord configuration file/priority to `50-rsyslogd-wrapper.conf`/`50`.
- Updates haproxy-wrapper supervisord configuration file/priority to `90-haproxy-wrapper.conf`/`90`.
- Updates docker-compose example with redis session store replacing memcached for the apache-php service.
- Updates wrapper scripts timer to use UTC date timestamps.
- Updates default timeout values.
- Updates proxy (varnish) configuration examples to use uri balance algorithm.
- Updates proxy (varnish) configuration examples to use both backends (removes backup).
- Fixes docker host connection status check in Makefile.
- Fixes docker-compose examples; volume mount errors on CentOS host.
- Adds `inspect`, `reload` and `top` Makefile targets.
- Adds improved `clean` Makefile target; includes exited containers and dangling images.
- Adds `SYSTEM_TIMEZONE` handling to Makefile, scmi, systemd unit and docker-compose templates.
- Adds system time zone validation to healthcheck.
- Adds lock/state file to wrapper scripts.
- Adds maintenance feature controlled via acl file `/etc/haproxy/maint-blacklist.acl`.
- Adds maintenance whitelist via acl file `/etc/haproxy/maint-whitelist.acl`.
- Removes support for long image tags (i.e. centos-7-2.x.x).
- Removes `dontlog-normal` from all configuration defaults and add to listeners.

### 2.2.0 - 2019-05-30

- Updates `haproxy18u` packages to 1.8.19-1.
- Updates source image to [2.5.1](https://github.com/jdeathe/centos-ssh/releases/tag/2.5.1).
- Updates and restructures Dockerfile.
- Updates container naming conventions and readability of `Makefile`.
- Updates healthcheck retries to 4.
- Updates docker-compose configuration examples.
- Updates default tls/ssl certificate name from `localhost.localdomain.crt` to `localhost.crt`.
- Fixes issue with unexpected published port in run templates when `DOCKER_PORT_MAP_TCP_80` or `DOCKER_PORT_MAP_TCP_443` is set to an empty string or 0.
- Fixes binary paths in systemd unit files for compatibility with both EL and Ubuntu hosts.
- Fixes environment variable name typo in README for `HAPROXY_CONFIG`.
- Adds port incrementation to Makefile's run template for container names with an instance suffix.
- Adds placeholder replacement of `RELEASE_VERSION` docker argument to systemd service unit template.
- Adds improvement to pull logic in systemd unit install template.
- Adds `SSH_AUTOSTART_SUPERVISOR_STDOUT` with a value "false", disabling startup of `supervisor_stdout`.
- Adds improved logging output.
- Adds consideration for event lag into test cases for unhealthy health_status events.
- Adds error messages to healthcheck script and includes supervisord check.
- Adds improved `healtchcheck`, `haproxy-wrapper` and `rsyslogd-wrapper` scripts.
- Adds improved lock/state file implementation in wrapper scripts.
- Adds config path and tls/ssl certificate fingerprint to `haproxy-wrapper` Details output.
- Adds support for hitless reload via `haproxy-wrapper`.
- Adds configuration of Apache certificate via `APACHE_SSL_CERTIFICATE` in `.env` for the tcp example.
- Adds SNI forwarding in the TLS/SSL tcp example configuration.
- Adds `/status` (`monitor-uri`) endpoints and custom error responses to http example configuration.
- Adds `socat` package to the image to allow for non-interactive HAProxy CLI usage.
- Removes use of `/etc/services-config` paths.
- Removes the unused group element from the default container name.
- Removes the node element from the default container name.
- Removes unused environment variables from Makefile and scmi configuration.
- Removes X-Fleet section from etcd register template unit-file.
- Removes use of `stick-table` in `haproxy-tcp.cfg` as it should not be necessary for web backends that support shared persistence/session data stores.

### 2.1.1 - 2018-12-27

- Updates `haproxy18u` packages to 1.8.14-1.
- Updates `rsyslog` packages to 8.24.0-34.
- Updates source image to [2.4.1](https://github.com/jdeathe/centos-ssh/releases/tag/2.4.1).
- Updates image versions in tests and docker-compose examples.
- Adds required `--sysctl` settings to docker run templates.
- Fixes haproxy-wrapper return error and normalises scripts.

test/shpec/operation_shpec.sh
readonly STARTUP_TIME=2
readonly TEST_DIRECTORY="test"

# These should ideally be a static value but hosts might be using this port so 
# need to allow for alternatives.
DOCKER_PORT_MAP_TCP_80="${DOCKER_PORT_MAP_TCP_80:-80}"
DOCKER_PORT_MAP_TCP_443="${DOCKER_PORT_MAP_TCP_443:-443}"

function __destroy ()
{
	local -r backend_alias_1="httpd_1"
	local -r backend_alias_2="httpd_2"
	local -r backend_name_1="apache-php.1"
	local -r backend_name_2="apache-php.2"
	local -r backend_network="bridge_t1"

	# Destroy the backend
	__terminate_container \
		${backend_name_1} \
	&> /dev/null

	__terminate_container \
		${backend_name_2} \
	&> /dev/null

	# Destroy the bridge network
	if [[ -n $(docker network ls -q -f name="${backend_network}") ]]; then
		docker network rm \
			${backend_network} \
		&> /dev/null
	fi
}

function __get_container_port ()
{
	local container="${1:-}"
	local port="${2:-}"
	local value=""

	value="$(
		docker port \
			${container} \
			${port}
	)"
	value=${value##*:}

	printf -- \
		'%s' \
		"${value}"
}

# container - Docker container name.
# counter - Timeout counter in seconds.
# process_pattern - Regular expression pattern used to match running process.
# ready_test - Command used to test if the service is ready.
function __is_container_ready ()
{
	local container="${1:-}"
	local counter=$(
		awk \
			-v seconds="${2:-10}" \
			'BEGIN { print 10 * seconds; }'
	)
	local process_pattern="${3:-}"
	local ready_test="${4:-true}"

	until (( counter == 0 )); do
		sleep 0.1

		if docker exec ${container} \
			bash -c "ps axo command \
				| grep -qE \"${process_pattern}\" \
				&& eval \"${ready_test}\"" \
			&> /dev/null
		then
			break
		fi

		(( counter -= 1 ))
	done

	if (( counter == 0 )); then
		return 1
	fi

	return 0
}

function __setup ()
{
	local -r backend_alias_1="httpd_1"
	local -r backend_alias_2="httpd_2"
	local -r backend_name_1="apache-php.1"
	local -r backend_name_2="apache-php.2"
	local -r backend_network="bridge_t1"
	local -r backend_release="1.13.3"

	# Create the bridge network
	if [[ -z $(docker network ls -q -f name="${backend_network}") ]]; then
		docker network create \
			--driver bridge \
			${backend_network} \
		&> /dev/null
	fi

	# Create the backend container
	__terminate_container \
		${backend_name_1} \
	&> /dev/null

	docker run \
		--detach \
		--name ${backend_name_1} \
		--env "APACHE_SERVER_NAME=www.app.local" \
		--network ${backend_network} \
		--network-alias ${backend_alias_1} \
		--volume ${PWD}/test/fixture/apache/var/www/public_html:/opt/app/public_html:ro \
		jdeathe/centos-ssh-apache-php:${backend_release} \
	&> /dev/null

	__terminate_container \
		${backend_name_2} \
	&> /dev/null

	docker run \
		--detach \
		--name ${backend_name_2} \
		--env "APACHE_SERVER_NAME=www.app.local" \
		--network ${backend_network} \
		--network-alias ${backend_alias_2} \
		--volume ${PWD}/test/fixture/apache/var/www/public_html:/opt/app/public_html:ro \
		jdeathe/centos-ssh-apache-php:${backend_release} \
	&> /dev/null

	# Generate a self-signed certificate
	openssl req \
		-x509 \
		-sha256 \
		-nodes \
		-newkey rsa:2048 \
		-days 365 \
		-subj "/CN=www.app.local" \
		-keyout /tmp/www.app.local.pem \
		-out /tmp/www.app.local.pem \
	&> /dev/null
}

# Custom shpec matcher
# Match a string with an Extended Regular Expression pattern.
function __shpec_matcher_egrep ()
{
	local pattern="${2:-}"
	local string="${1:-}"

	printf -- \
		'%s' \
		"${string}" \
	| grep -qE -- \
		"${pattern}" \
		-

	assert equal \
		"${?}" \
		0
}

function __terminate_container ()
{
	local container="${1}"

	if docker ps -aq \
		--filter "name=${container}" \
		--filter "status=paused" &> /dev/null; then
		docker unpause ${container} &> /dev/null
	fi

	if docker ps -aq \
		--filter "name=${container}" \
		--filter "status=running" &> /dev/null; then
		docker stop ${container} &> /dev/null
	fi

	if docker ps -aq \
		--filter "name=${container}" &> /dev/null; then
		docker rm -vf ${container} &> /dev/null
	fi
}

function test_basic_operations ()
{
	local -r backend_hostname="localhost.localdomain"
	local -r backend_name_1="apache-php.1"
	local -r backend_name_2="apache-php.2"
	local -r backend_network="bridge_t1"
	local -r content="$(< test/fixture/apache/var/www/public_html/index.html)"

	local backend_content=""
	local backend_response_code=""
	local cli_server=""
	local cli_server_state="maint"
	local cli_servers="http/web_1 http/web_2 https/web_1 https/web_2"
	local container_port_80=""
	local container_port_443=""
	local logs_since=1
	local logs_timeout=60

	describe "Basic HAProxy operations"
		trap "__terminate_container haproxy.1 &> /dev/null; \
			__destroy; \
			exit 1" \
			INT TERM EXIT

		__terminate_container \
			haproxy.1 \
		&> /dev/null

		describe "Runs named container"
			docker run \
				--detach \
				--name haproxy.1 \
				--network ${backend_network} \
				--publish ${DOCKER_PORT_MAP_TCP_80}:80 \
				--publish ${DOCKER_PORT_MAP_TCP_443}:443 \
				jdeathe/centos-ssh-haproxy:latest \
			&> /dev/null
	
			it "Can publish ${DOCKER_PORT_MAP_TCP_80}:80."
				container_port_80="$(
					__get_container_port \
						haproxy.1 \
						80/tcp
				)"

				if [[ ${DOCKER_PORT_MAP_TCP_80} == 0 ]] \
					|| [[ -z ${DOCKER_PORT_MAP_TCP_80} ]]; then
					assert gt \
						"${container_port_80}" \
						"30000"
				else
					assert equal \
						"${container_port_80}" \
						"${DOCKER_PORT_MAP_TCP_80}"
				fi
			end

			it "Can publish ${DOCKER_PORT_MAP_TCP_443}:443."
				container_port_443="$(
					__get_container_port \
						haproxy.1 \
						443/tcp
				)"

				if [[ ${DOCKER_PORT_MAP_TCP_443} == 0 ]] \
					|| [[ -z ${DOCKER_PORT_MAP_TCP_443} ]]; then
					assert gt \
						"${container_port_443}" \
						"30000"
				else
					assert equal \
						"${container_port_443}" \
						"${DOCKER_PORT_MAP_TCP_443}"
				fi
			end
		end

		if ! __is_container_ready \
			haproxy.1 \
			${STARTUP_TIME} \
			"/usr/sbin/haproxy " \
			"/usr/bin/healthcheck"
		then
			exit 1
		fi

		describe "HTTP requests"
			describe "Unencrypted response"
				it "Is unaltered."
					#TODO remove workaround added for https://github.com/jdeathe/centos-ssh-apache-php/issues/786
					backend_content="$(
						curl -s \
							-H "Host: ${backend_hostname}" \
							http://127.0.0.1:${container_port_80}/index.html
					)"

					assert equal \
						"${backend_content}" \
						"${content}"
				end
			end

			describe "Encrypted response"
				it "Is unaltered."
					#TODO remove workaround added for https://github.com/jdeathe/centos-ssh-apache-php/issues/786
					backend_content="$(
						curl -sk \
							-H "Host: ${backend_hostname}" \
							https://127.0.0.1:${container_port_443}/index.html
					)"

					assert equal \
						"${backend_content}" \
						"${content}"
				end
			end
		end

		describe "Monitor URI"
			describe "Backend up"
				describe "Unencrypted response"
					it "Is 200 OK."
						backend_response_code="$(
							curl -sI \
								-o /dev/null \
								-w "%{http_code}" \
								-H "Host: ${backend_hostname}" \
								http://127.0.0.1:${container_port_80}/status
						)"

						backend_content="$(
							curl -s \
								-H "Host: ${backend_hostname}" \
								http://127.0.0.1:${container_port_80}/status
						)"

						assert equal \
							"${backend_response_code}:${backend_content}" \
							"200:OK"
					end
				end

				describe "Encrypted response"
					it "Is 200 OK."
						backend_response_code="$(
							curl -skI \
								-o /dev/null \
								-w "%{http_code}" \
								-H "Host: ${backend_hostname}" \
								https://127.0.0.1:${container_port_443}/status
						)"

						backend_content="$(
							curl -sk \
								-H "Host: ${backend_hostname}" \
								https://127.0.0.1:${container_port_443}/status
						)"

						assert equal \
							"${backend_response_code}:${backend_content}" \
							"200:OK"
					end
				end
			end

			describe "Backend down"
				# Put backend servers into maintenance state
				cli_server_state="maint"
				for cli_server in ${cli_servers}
				do
					docker exec -i \
						haproxy.1 \
						socat - UNIX:/var/lib/haproxy/stats-1 \
						<<< "set server ${cli_server} state ${cli_server_state}" \
					&> /dev/null
				done

				describe "Unencrypted response"
					it "Is 503 Service Unavailable."
						backend_response_code="$(
							curl -sI \
								-o /dev/null \
								-w "%{http_code}" \
								-H "Host: ${backend_hostname}" \
								http://127.0.0.1:${container_port_80}/status
						)"

						backend_content="$(
							curl -s \
								-H "Host: ${backend_hostname}" \
								http://127.0.0.1:${container_port_80}/status
						)"

						assert equal \
							"${backend_response_code}:${backend_content}" \
							"503:Service Unavailable"
					end
				end

				describe "Encrypted response"
					it "Is 503 Service Unavailable."
						backend_response_code="$(
							curl -skI \
								-o /dev/null \
								-w "%{http_code}" \
								-H "Host: ${backend_hostname}" \
								https://127.0.0.1:${container_port_443}/status
						)"

						backend_content="$(
							curl -sk \
								-H "Host: ${backend_hostname}" \
								https://127.0.0.1:${container_port_443}/status
						)"

						assert equal \
							"${backend_response_code}:${backend_content}" \
							"503:Service Unavailable"
					end
				end
			end
		end

		__terminate_container \
			haproxy.1 \
		&> /dev/null

		trap - \
			INT TERM EXIT
	end
}

function test_custom_configuration ()
{
	local -r backend_hostname="www.app.local"
	local -r backend_network="bridge_t1"
	local -r content="$(< test/fixture/apache/var/www/public_html/index.html)"

	local backend_content=""
	local certificate_fingerprint_file=""
	local certificate_fingerprint_server=""
	local certificate_pem_base64=""
	local container_port_80=""
	local container_port_443=""

	describe "Customised HAProxy operations"
		trap "__terminate_container haproxy.1 &> /dev/null; \
			__destroy; \
			exit 1" \
			INT TERM EXIT

		describe "SSL/TLS"
			describe "Static certificate."
				if [[ -s /tmp/www.app.local.pem ]]; then
					certificate_fingerprint_file="$(
						cat \
							/tmp/www.app.local.pem \
						| sed \
							-n \
							-e '/^-----BEGIN CERTIFICATE-----$/,/^-----END CERTIFICATE-----$/p' \
						| openssl x509 \
							-fingerprint \
							-noout \
						| sed \
							-e 's~SHA1 Fingerprint=~~'
					)"

					if [[ $(uname) == "Darwin" ]]; then
						certificate_pem_base64="$(
							base64 \
								-i /tmp/www.app.local.pem
						)"
					else
						certificate_pem_base64="$(
							base64 \
								-w 0 \
								-i /tmp/www.app.local.pem
						)"
					fi

					printf \
						-- '%s' \
						"${certificate_pem_base64}" \
					> /tmp/www.app.local.txt
				fi

				it "Sets from base64 encoded value."
					__terminate_container \
						haproxy.1 \
					&> /dev/null

					docker run \
						--detach \
						--name haproxy.1 \
						--env HAPROXY_SSL_CERTIFICATE="${certificate_pem_base64}" \
						--network ${backend_network} \
						--publish ${DOCKER_PORT_MAP_TCP_80}:80 \
						--publish ${DOCKER_PORT_MAP_TCP_443}:443 \
						--volume /tmp:/var/run/tmp:ro \
						jdeathe/centos-ssh-haproxy:latest \
					&> /dev/null

					container_port_443="$(
						__get_container_port \
							haproxy.1 \
							443/tcp
					)"

					if ! __is_container_ready \
						haproxy.1 \
						${STARTUP_TIME} \
						"/usr/sbin/haproxy " \
						"/usr/bin/healthcheck"
					then
						exit 1
					fi

					certificate_fingerprint_server="$(
						echo -n \
						| openssl s_client \
							-connect 127.0.0.1:${container_port_443} \
							-CAfile /tmp/www.app.local.pem \
							-nbio \
							2>&1 \
						| sed \
							-n \
							-e '/^-----BEGIN CERTIFICATE-----$/,/^-----END CERTIFICATE-----$/p' \
						| openssl \
							x509 \
							-fingerprint \
							-noout \
						| sed \
							-e 's~SHA1 Fingerprint=~~'
					)"

					assert equal \
						"${certificate_fingerprint_server}" \
						"${certificate_fingerprint_file}"
				end

				it "Sets from file path value."
					__terminate_container \
						haproxy.1 \
					&> /dev/null

					docker run \
						--detach \
						--name haproxy.1 \
						--env HAPROXY_SSL_CERTIFICATE="/var/run/tmp/www.app.local.pem" \
						--network ${backend_network} \
						--publish ${DOCKER_PORT_MAP_TCP_80}:80 \
						--publish ${DOCKER_PORT_MAP_TCP_443}:443 \
						--volume /tmp:/var/run/tmp:ro \
						jdeathe/centos-ssh-haproxy:latest \
					&> /dev/null

					container_port_443="$(
						__get_container_port \
							haproxy.1 \
							443/tcp
					)"

					if ! __is_container_ready \
						haproxy.1 \
						${STARTUP_TIME} \
						"/usr/sbin/haproxy " \
						"/usr/bin/healthcheck"
					then
						exit 1
					fi

					certificate_fingerprint_server="$(
						echo -n \
						| openssl s_client \
							-connect 127.0.0.1:${container_port_443} \
							-CAfile /tmp/www.app.local.pem \
							-nbio \
							2>&1 \
						| sed \
							-n \
							-e '/^-----BEGIN CERTIFICATE-----$/,/^-----END CERTIFICATE-----$/p' \
						| openssl \
							x509 \
							-fingerprint \
							-noout \
						| sed \
							-e 's~SHA1 Fingerprint=~~'
					)"

					assert equal \
						"${certificate_fingerprint_server}" \
						"${certificate_fingerprint_file}"
				end

				it "Sets from file path with base64 encoded content."
					__terminate_container \
						haproxy.1 \
					&> /dev/null

					docker run \
						--detach \
						--name haproxy.1 \
						--env HAPROXY_SSL_CERTIFICATE="/var/run/tmp/www.app.local.txt" \
						--network ${backend_network} \
						--publish ${DOCKER_PORT_MAP_TCP_80}:80 \
						--publish ${DOCKER_PORT_MAP_TCP_443}:443 \
						--volume /tmp:/var/run/tmp:ro \
						jdeathe/centos-ssh-haproxy:latest \
					&> /dev/null

					container_port_443="$(
						__get_container_port \
							haproxy.1 \
							443/tcp
					)"

					if ! __is_container_ready \
						haproxy.1 \
						${STARTUP_TIME} \
						"/usr/sbin/haproxy " \
						"/usr/bin/healthcheck"
					then
						exit 1
					fi

					certificate_fingerprint_server="$(
						echo -n \
						| openssl s_client \
							-connect 127.0.0.1:${container_port_443} \
							-CAfile /tmp/www.app.local.pem \
							-nbio \
							2>&1 \
						| sed \
							-n \
							-e '/^-----BEGIN CERTIFICATE-----$/,/^-----END CERTIFICATE-----$/p' \
						| openssl \
							x509 \
							-fingerprint \
							-noout \
						| sed \
							-e 's~SHA1 Fingerprint=~~'
					)"

					assert equal \
						"${certificate_fingerprint_server}" \
						"${certificate_fingerprint_file}"
				end
			end
		end

		__terminate_container \
			haproxy.1 \
		&> /dev/null

		trap - \
			INT TERM EXIT
	end
}

function test_healthcheck ()
{
	local -r backend_network="bridge_t1"
	local -r event_lag_seconds=2
	local -r interval_seconds=1
	local -r retries=4
	local container_id
	local events_since_timestamp
	local health_status

	trap "__terminate_container haproxy.1 &> /dev/null; \
		__destroy; \
		exit 1" \
		INT TERM EXIT

	describe "Healthcheck"
		describe "Default configuration"
			__terminate_container \
				haproxy.1 \
			&> /dev/null

			docker run \
				--detach \
				--name haproxy.1 \
				--network ${backend_network} \
				jdeathe/centos-ssh-haproxy:latest \
			&> /dev/null

			events_since_timestamp="$(
				date +%s
			)"

			container_id="$(
				docker ps \
					--quiet \
					--filter "name=haproxy.1"
			)"

			it "Returns a valid status on starting."
				health_status="$(
					docker inspect \
						--format='{{json .State.Health.Status}}' \
						haproxy.1
				)"

				assert __shpec_matcher_egrep \
					"${health_status}" \
					"\"(starting|healthy|unhealthy)\""
			end

			it "Returns healthy after startup."
				events_timeout="$(
					awk \
						-v event_lag="${event_lag_seconds}" \
						-v interval="${interval_seconds}" \
						-v startup_time="${STARTUP_TIME}" \
						'BEGIN { print event_lag + startup_time + interval; }'
				)"

				health_status="$(
					test/health_status \
						--container="${container_id}" \
						--since="${events_since_timestamp}" \
						--timeout="${events_timeout}" \
						--monochrome \
					2>&1
				)"

				assert equal \
					"${health_status}" \
					"✓ healthy"
			end

			it "Returns unhealthy on failure."
				docker exec -t \
					haproxy.1 \
					bash -c "mv \
						/usr/sbin/haproxy \
						/usr/sbin/haproxy2" \
				&& docker exec -t \
					haproxy.1 \
					bash -c "if [[ -n \$(pgrep -f '^/usr/sbin/haproxy ') ]]; then \
						kill -9 \$(pgrep -f '^/usr/sbin/haproxy ')
					fi"

				events_since_timestamp="$(
					date +%s
				)"

				events_timeout="$(
					awk \
						-v event_lag="${event_lag_seconds}" \
						-v interval="${interval_seconds}" \
						-v retries="${retries}" \
						'BEGIN { print (2 * event_lag) + (interval * retries); }'
				)"

				health_status="$(
					test/health_status \
						--container="${container_id}" \
						--since="$(( ${event_lag_seconds} + ${events_since_timestamp} ))" \
						--timeout="${events_timeout}" \
						--monochrome \
					2>&1
				)"

				assert equal \
					"${health_status}" \
					"✗ unhealthy"
			end

			__terminate_container \
				haproxy.1 \
			&> /dev/null
		end
	end

	trap - \
		INT TERM EXIT
}

if [[ ! -d ${TEST_DIRECTORY} ]]; then
	printf -- \
		"ERROR: Please run from the project root.\n" \
		>&2
	exit 1
fi

describe "jdeathe/centos-ssh-haproxy:latest"
	__destroy
	__setup
	test_basic_operations
	test_custom_configuration
	test_healthcheck
	__destroy
end


src/opt/scmi/environment.sh
# ------------------------------------------------------------------------------
# Constants
# ------------------------------------------------------------------------------
readonly DOCKER_IMAGE_NAME=centos-ssh-haproxy
readonly DOCKER_IMAGE_RELEASE_TAG_PATTERN='^[1-2]\.[0-9]+\.[0-9]+$'
readonly DOCKER_IMAGE_TAG_PATTERN='^(latest|[1-2]\.[0-9]+\.[0-9]+)$'
readonly DOCKER_USER=jdeathe

# ------------------------------------------------------------------------------
# Variables
# ------------------------------------------------------------------------------
DIST_PATH="${DIST_PATH:-./dist}"
DOCKER_CONTAINER_OPTS="${DOCKER_CONTAINER_OPTS:-}"
DOCKER_IMAGE_TAG="${DOCKER_IMAGE_TAG:-latest}"
DOCKER_NAME="${DOCKER_NAME:-haproxy.1}"
DOCKER_PORT_MAP_TCP_80="${DOCKER_PORT_MAP_TCP_80:-80}"
DOCKER_PORT_MAP_TCP_443="${DOCKER_PORT_MAP_TCP_443:-443}"
DOCKER_RESTART_POLICY="${DOCKER_RESTART_POLICY:-always}"
NO_CACHE="${NO_CACHE:-false}"
REGISTER_ETCD_PARAMETERS="${REGISTER_ETCD_PARAMETERS:-}"
REGISTER_TTL="${REGISTER_TTL:-60}"
REGISTER_UPDATE_INTERVAL="${REGISTER_UPDATE_INTERVAL:-55}"
STARTUP_TIME="${STARTUP_TIME:-2}"
SYSCTL_NET_CORE_SOMAXCONN="${SYSCTL_NET_CORE_SOMAXCONN:-32768}"
SYSCTL_NET_IPV4_IP_LOCAL_PORT_RANGE="${SYSCTL_NET_IPV4_IP_LOCAL_PORT_RANGE:-"1024 65535"}"
SYSCTL_NET_IPV4_ROUTE_FLUSH="${SYSCTL_NET_IPV4_ROUTE_FLUSH:-1}"
ULIMIT_MEMLOCK="${ULIMIT_MEMLOCK:-82000}"
ULIMIT_NOFILE="${ULIMIT_NOFILE:-131072}"
ULIMIT_NPROC="${ULIMIT_NPROC:-9223372036854775807}"

# ------------------------------------------------------------------------------
# Application container configuration
# ------------------------------------------------------------------------------
HAPROXY_CONFIG="${HAPROXY_CONFIG:-/etc/haproxy/haproxy.cfg}"
HAPROXY_HOST_NAMES="${HAPROXY_HOST_NAMES:-localhost.localdomain}"
HAPROXY_SSL_CERTIFICATE="${HAPROXY_SSL_CERTIFICATE:-}"
SYSTEM_TIMEZONE="${SYSTEM_TIMEZONE:-UTC}"
